verification.go

```diff
@@ -0,0 +1,101 @@
+package neutrino
+
+import (
+	"fmt"
+
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcutil"
+	"github.com/btcsuite/btcutil/gcs"
+	"github.com/btcsuite/btcutil/gcs/builder"
+)
+
+// VerifyBasicBlockFilter asserts that a given block filter was constructed
+// correctly and according to the rules of BIP-0158 to contain both the output's
+// pk scripts as well as the pk scripts the inputs are spending.
+func VerifyBasicBlockFilter(filter *gcs.Filter, block *btcutil.Block) error {
+	key := builder.DeriveKey(block.Hash())
+
+	for idx, tx := range block.Transactions() {
+		// Skip coinbase transaction.
+		if idx == 0 {
+			continue
+		}
+
+		// Check outputs first.
+		for outIdx, out := range tx.MsgTx().TxOut {
+			match, err := filter.MatchAny(
+				key, [][]byte{out.PkScript},
+			)
+			if err != nil {
+				return fmt.Errorf("error validating block %v "+
+					"outpoint %v:%d script %x: %v",
+					block.Hash(), tx.Hash(), outIdx,
+					out.PkScript, err)
+			}
+
+			if !match {
+				return fmt.Errorf("filter for block %v is "+
+					"invalid, outpoint %v:%d script %x "+
+					"wasn't matched by filter",
+					block.Hash(), tx.Hash(), outIdx,
+					out.PkScript)
+			}
+		}
+
+		// Now we can go through all inputs and check that the filter
+		// also included any pk scripts of the outputs being _spent_.
+		// We can do this for witness items since the witness always
+		// contains the full script as the last element on the stack.
+		for inIdx, in := range tx.MsgTx().TxIn {
+			script, err := txscript.ComputePkScript(
```

Comment:
Note that as is, this doesn't cover nested P2SH, as the logic in `ComputePkScript` branches based on the length of the witness and sigScript. We can bridge this by checking if _both_ are active, then we'd compute the witness program based on the witness script, then compute the P2SH script of that and assert everything matches up. cc @wpaulino

Reply:
Actually, I think it's fine as is, since we'll return the P2SH script here which is what should be inserted into the filter...